### ISSUE TITLE
CORS ACAO error with docker container name in REGISTRY_URL, access denied to registry

### ISSUE THREAD

**Issue by gyroplast** [Joxit/docker-registry-ui#220]
Greetings!

There seems to be an issue with name resolution in HTTP requests to the configured `REGISTRY_URL` when using docker container names. 

## Bug description

An insecure registry running as a docker container cannot be accessed by the docker-registry-ui running in another container when I supply the registry container name in the `REGISTRY_URL` environment variable (`http://myreg:5000`), ostensibly due to a blocked cross-origin request, despite the registry being configured accordingly and verified to respond with `Access-Control-Allow-Origin: *` in the headers.

The GUI displays red error message when directing browser to `http://127.0.0.1:10012`: 
> An error occured: Check your connection and your registry must have `Access-Control-Allow-Origin` header set to `http://127.0.0.1:10012`

Firefox' console prints accordingly:
> Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://myreg:5000/v2/_catalog?n=100000. (Reason: CORS request did not succeed).

Both containers, `myreg` and `registry_ui`, are attached to the same docker network.
Supplying the docker container IPv4 address (`http://172.25.0.2:5000`), or an "external" hostname outside the docker networking name resolution space, f. ex. the hostname of the machine the containers are running on (`http://vixen:5000`), instead of the container name in `REGISTRY_URL` works fine.
Using the docker container ID or "fully-qualified" container name with network suffix as hostnames in the `REGISTRY_URL` also fails.
All these variants _are_ correctly resolved, and IP-reachable from the `registry_ui` container.
Running f. ex. `docker exec registry_ui curl -v http://myreg:5000/v2/_catalog?n=100000` succeeds and shows the `Access-Control-Allow-Origin: *` header being set in the response.

I could reproduce the exact same issue independent of creating both containers directly, using docker-compose, or even deploying as a stack.

It seems that only the JavaScript request to the registry exhibits this issue, but no other application running in the context of the `registry_ui` container. Whilst I cannot rule out a specific misunderstanding on my side regarding docker networking, according to my understanding my docker setup as detailed below is sound and should work as intended.

## How to Reproduce

I'll describe reproduction with `docker-compose` as this is likely to be easiest to get up and running. The problem was encountered with an equivalent stack deployment and direct container running with the docker CLI as well, though.

My docker-compose file:
```yml
version: '3.5'

services:
  myreg:
    image: registry:2
    ports:
    - 5000:5000
    environment:
    - "REGISTRY_HTTP_HEADERS_Access-Control-Allow-Origin=['*']"

  registry_ui:
    image: joxit/docker-registry-ui:latest
    ports:
    - 10012:80
    environment:
    - "REGISTRY_URL=http://myreg:5000"
```

```shell
[deh@vixen tmp]$ docker-compose up -d
[+] Running 3/3                                                                                
 ⠿ Network tmp_default          Created   0.2s
 ⠿ Container tmp-myreg-1        Started   1.0s
 ⠿ Container tmp-registry_ui-1  Started   1.0s

[deh@vixen tmp]$ docker-compose logs
tmp-myreg-1  | time="2021-11-05T14:47:47.765808223Z" level=warning msg="No HTTP secret provided - generated random secret. This may cause problems with uploads if multiple registries are behind a load-balancer. To provide a shared secret, fill in http.secret in the configuration file or set the REGISTRY_HTTP_SECRET environment variable." go.version=go1.11.2 instance.id=e0e1751b-c435-42b0-b207-81bd2b9d26d6 service=registry version=v2.7.1 
tmp-myreg-1  | time="2021-11-05T14:47:47.765838827Z" level=info msg="Starting upload purge in 10m0s" go.version=go1.11.2 instance.id=e0e1751b-c435-42b0-b207-81bd2b9d26d6 service=registry version=v2.7.1 
tmp-myreg-1  | time="2021-11-05T14:47:47.765868874Z" level=info msg="redis not configured" go.version=go1.11.2 instance.id=e0e1751b-c435-42b0-b207-81bd2b9d26d6 service=registry version=v2.7.1 
tmp-myreg-1  | time="2021-11-05T14:47:47.772645189Z" level=info msg="using inmemory blob descriptor cache" go.version=go1.11.2 instance.id=e0e1751b-c435-42b0-b207-81bd2b9d26d6 service=registry version=v2.7.1 
tmp-myreg-1  | time="2021-11-05T14:47:47.772747653Z" level=info msg="listening on [::]:5000" go.version=go1.11.2 instance.id=e0e1751b-c435-42b0-b207-81bd2b9d26d6 service=registry version=v2.7.1 
tmp-registry_ui-1  | /docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
tmp-registry_ui-1  | /docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
tmp-registry_ui-1  | /docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh
tmp-registry_ui-1  | 10-listen-on-ipv6-by-default.sh: info: Getting the checksum of /etc/nginx/conf.d/default.conf
tmp-registry_ui-1  | 10-listen-on-ipv6-by-default.sh: info: /etc/nginx/conf.d/default.conf differs from the packaged version
tmp-registry_ui-1  | /docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh
tmp-registry_ui-1  | /docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh
tmp-registry_ui-1  | /docker-entrypoint.sh: Launching /docker-entrypoint.d/90-docker-registry-ui.sh
tmp-registry_ui-1  | /docker-entrypoint.sh: Configuration complete; ready for start up
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: using the "epoll" event method
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: nginx/1.21.3
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: built by gcc 10.3.1 20210424 (Alpine 10.3.1_git20210424) 
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: OS: Linux 5.14.14-hardened1-1-hardened
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker processes
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 40
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 41
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 42
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 43
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 44
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 45
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 46
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 47
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 48
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 49
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 50
tmp-registry_ui-1  | 2021/11/05 14:47:47 [notice] 1#1: start worker process 51

[deh@vixen tmp]$ docker-compose exec registry_ui curl -v http://myreg:5000/v2/_catalog?n=100000
*   Trying 172.25.0.2:5000...
* Connected to myreg (172.25.0.2) port 5000 (#0)
> GET /v2/_catalog?n=100000 HTTP/1.1
> Host: myreg:5000
> User-Agent: curl/7.78.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Access-Control-Allow-Origin: *
< Content-Type: application/json; charset=utf-8
< Docker-Distribution-Api-Version: registry/2.0
< X-Content-Type-Options: nosniff
< Date: Fri, 05 Nov 2021 14:58:03 GMT
< Content-Length: 20
< 
{"repositories":[]}
* Connection #0 to host myreg left intact

```

## Expected behavior

After starting the containers, opening http://127.0.0.1:10012 in a browser should display an empty registry overview without error. Instead, I am presented with a CORS error:

![Screenshot_2021-11-05_16-00-26_error](https://user-images.githubusercontent.com/2656777/140532924-11a80aa3-be41-4c81-995f-b8769423cc08.png)

Changing the `REGISTRY_URL` to point to the registry container's IP address works as expected:

```
[deh@vixen tmp]$ docker-compose stop registry_ui
[+] Running 1/1
 ⠿ Container tmp-registry_ui-1  Stopped    0.7s

[deh@vixen tmp]$ docker-compose rm -f registry_ui
Going to remove tmp-registry_ui-1
[+] Running 1/0
 ⠿ Container tmp-registry_ui-1  Removed    0.1s

[deh@vixen tmp]$ sed -i 's|REGISTRY_URL=http://myreg:5000|REGISTRY_URL=http://172.25.0.2:5000|' docker-compose.yml 

[deh@vixen tmp]$ docker-compose up -d
[+] Running 2/2
 ⠿ Container tmp-registry_ui-1  Started    0.6s
 ⠿ Container tmp-myreg-1        Running    0.0s
```

Reloading the page in the browser then looks perfectly fine:

![Screenshot_2021-11-05_16-17-31_success](https://user-images.githubusercontent.com/2656777/140534686-2eb396bf-7d59-4968-900f-b0e7c65640b1.png)
 
## System information

- OS: Arch Linux on native x86_64
- Docker registry UI:
    - Version: 2.0.9
    - Server: docker
    - Docker version: 20.10.9
    - Docker registry ui tag: latest (0e6d7078a460)
    - OS/Arch: linux/amd64
    - Tools: local docker CLI, docker-compose, and docker swarm

## Additional context

As stated before, I only encounter this problem when using any container name variant in the `REGISTRY_URL` that should be resolved just as well as any other hostname lookup. Using a fully-qualified external address (i. e. vulpes.veloxis.de:5000) where I ran the exact same registry container temporarily worked flawlessly, so did using a hostname-only pointing to my docker host's wifi IP address (vixen:5000 at 192.168.30.1).

Something is messy, but _only_ when referencing the docker container in a somewhat isolated service/docker network setup, which I consider best practice when creating docker stacks.

Am I doing or understanding something fundamentally wrong in respect to docker networking, or is there something really weird going on with the used JavaScript module for HTTP requests in this application?

**Comment by Joxit** [comment]
Hello, thank your for using my project :smile: 

TLDR: this is not a CORS issue but a `ERR_NAME_NOT_RESOLVED`, Firefox does not differentiate network issues and CORS, all is CORS for Firefox.... Your browser has confused you... :confused: The correct test to reproduce this on your computer is `curl -v http://myreg:5000/v2/_catalog?n=100000` on you computer and **NOT in a container**.

I did a small diagram for you: 
![explain-docker-registry drawio](https://user-images.githubusercontent.com/5153882/140613512-aef3b5e1-dfb5-446a-b2c5-971ef9299776.png)

When you use the option `REGISTRY_URL` you tell to the registry UI how to communicate with your registry via global network. In this example there are **no communications** between the UI and the registry itself.

You with your browser does all the communications, the first one to get all HTML + JS from docker registry UI, then all images/tags from the docker registry.

This means that your browser **MUST** know `myreg`, but this is available only in the docker network and not on your computer... That's why `vixen` and the IP worked fine, vixen is configured by your PC, and the IP is not a hostname :smile: 

So resolutions are: 
- `REGISTRY_URL=http://127.0.0.1:5000`
- use `NGINX_PROXY_PASS_URL` instead

New diagram to understand how work `REGISTRY_URL` VS `NGINX_PROXY_PASS_URL`
![explain-docker-registry-ui drawio](https://user-images.githubusercontent.com/5153882/140614332-7a211cc4-1167-4b91-991f-51b2b307cb4e.png)

I hope this helped. Tell me the diagram is clear enough, I could add it to the documentation.

**Comment by gyroplast** [comment]
Thank you for the effort, this makes perfect sense. I was under the impression that the registry-ui app itself was communicating with the registry "server-sided", and I didn't even entertain the option of a client-side connection from my browser to the registry via JS.

The information to set `REGISTRY_URL` from the POV of the client instead of the registry-ui application was the missing puzzle piece for me. It might be helpful to clarify this in the documentation eventually, but for what it's worth, my issue is resolved. Not a bug after all. ;)

Thanks again for the nifty little tool, it's doing a really good and simple job!

**Comment by Joxit** [comment]
It's a pleasure 😊

Yes, every options are from the point of view of the client except those stating with NGINX_*
